torchtitan/experiments/deepseek_v3/model.py

```diff
@@ -706,23 +730,156 @@ def moe_forward(self, x, topk_ids, topk_weight):
+    def _run_group_gemm(self, contig_tokens, m_sizes, m_offsets, permuted_indices):
+        if self.group_mm in ["torch", "torchao"]:
+            # Get weights
+            w_gate = self.get_parameter("gate_proj_weight")
+            w_up = self.get_parameter("up_proj_weight")
+            w_down = self.get_parameter("down_proj_weight")
+
+            # Run first two GEMMs (gate and up projections)
+            if self.group_mm == "torchao":
+                gate_proj = grouped_gemm_forward(contig_tokens, w_gate, m_sizes)
+                up_proj = grouped_gemm_forward(contig_tokens, w_up, m_sizes)
+            else:  # "torch"
+                gate_proj = torch._grouped_mm(
+                    contig_tokens,
+                    w_gate.transpose(-2, -1),
+                    m_offsets,
+                    out_dtype=torch.bfloat16,
+                )
+                up_proj = torch._grouped_mm(
+                    contig_tokens,
+                    w_up.transpose(-2, -1),
+                    m_offsets,
+                    out_dtype=torch.bfloat16,
+                )
+
+            # Apply activation
+            hidden_outputs = MLP.act_fn(gate_proj) * up_proj
+
+            # Run the third GEMM (down projection)
+            if self.group_mm == "torchao":
+                hidden_outputs = grouped_gemm_forward(hidden_outputs, w_down, m_sizes)
+            else:  # "torch"
+                hidden_outputs = torch._grouped_mm(
+                    hidden_outputs,
+                    w_down.transpose(-2, -1),
+                    m_offsets,
+                    out_dtype=torch.bfloat16,
+                )
+
+            return hidden_outputs
+
+        elif self.group_mm == "ds":
+            valid_tokens = contig_tokens[: m_offsets[-1]]
+            m_indices = dsgemm_utils.create_indices_from_offsets_nosync(m_offsets)
+
+            # Get expert weights for all projections
+            gate_proj_weight_fp8 = self.get_parameter("gate_proj_fp8")
+            gate_proj_scales = self.get_parameter("gate_proj_scales")
+            up_proj_weight_fp8 = self.get_parameter("up_proj_fp8")
+            up_proj_scales = self.get_parameter("up_proj_scales")
+            down_proj_weight_fp8 = self.get_parameter("down_proj_fp8")
+            down_proj_scales = self.get_parameter("down_proj_scales")
+
+            # Get dimensions
+            m_actual_tokens = valid_tokens.shape[0]
+            intermediate_size = self.get_parameter("gate_proj_weight").shape[1]
+            hidden_size = self.get_parameter("down_proj_weight").shape[1]
+
+            # Prepare input in FP8 format (shared by gate and up projections)
+            if self.use_triton_quant:
+                gate_up_input = dsgemm_kernels.groupwise_activation_quant(
+                    valid_tokens
+                )  # dsgemm_kernels.activation_quant_triton(valid_tokens)
+            else:
+                gate_up_input = dsgemm_utils.prepare_fp8_input(valid_tokens)
```

Comment:
What's the true meaning of this if-else? Just a selection of different impls? Or do they mean different quantization granularity?

Reply:
it's the same end result (1x128 quantization) - rather it's whether we are using our triton kernel(s) to do it or the dsgemm built in utils. 
I think this is now clarified via the comment add to the self.use_triton_quant during init, and also given that the triton kernels are faster longer term this will go away.  I wanted to keep it though to verify once we have training options for dsgemm.